### PR TITLE
Data structure: Linked list

### LINKED ISSUE #16: Collections: linked_list

### PR DESCRIPTION
**Pull request by delvor**
- cleaned typos  in heap_sort

- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [ ] PR has appropriate labels and milestones for easy identification
- [ ] PR it is assigned to the owner
- [ ] reviewers are assigned
- [ ] the PR contributes only one focused change
- [ ] the branch is up to date with `main`/`master`
- [ ] the code works when pulled and run locally
- [ ] all conflicts are resolved (if any)
- [ ] It is linked to an issue in the appropriate column of the project board
- [ ] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [ ] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [x] There is a module header
- [x] There is a module docstring
- [x] File names are in snake_case
- [ ] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [ ] Parameter description
- [ ] Return value description
- [ ] Include any assumptions
- [ ] Include 3 or more (passing!) doctests
- [ ] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [ ] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [x] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

### CODE REVIEW

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        """Creates empty linked list"""
+        self.__data: list[self.Node] = list()
+        self.__number_of_elements: int = 0
+
+    def empty(self) -> bool:
+        """Returns True if empty, False if not empty"""
+        return self.__number_of_elements == 0
+
+    def size(self) -> int:
+        """Returns True if empty, False if not empty"""
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:33]: looks like it should return size, and this docstrign is copied )

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        self.__number_of_elements += 1
+
+    def front(self) -> any:
+        """Returns first element"""
+        element: any = 0
+        if not self.empty():
+            element = self.__data[0].value
+        else:
+            element = None
+        return element
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:52]: can be done just without `element` with returns

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        """Returns last element"""
+        element: any = 0
+        if not self.empty():
+            element = self.__data[-1].value
+        else:
+            element = None
+        return element
+
+    def pop_front(self):
+        """Remove first element from list"""
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:64]: just a question - I thought .pop() function takes element from the end of list, remove it from list and **returns** ?
self.__data.pop(0) - will return data[0] before deleting it

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        return element
+
+    def back(self) -> any:
+        """Returns last element"""
+        element: any = 0
+        if not self.empty():
+            element = self.__data[-1].value
+        else:
+            element = None
+        return element
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:61]: but maybe the code looks better

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            pointer: self.Node = self.__data[counter]
+            
+            # Iterating thru nodes
+            while pointer.next != None:
+                # comparing values of nodes 
+                if pointer.value == element:
+
+                    if counter-1 > -1:
+                        #Change referense to next element
+                        self.__data[counter-1].next = pointer
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:100]: it is doing nothing, from reading code this like should looks like
  self.__data[counter-1].next = pointer.next

otherwise you leave reference in the [counter-1] element to _this_ node, which you will delete


or move this block after line 103 (after  pointer = pointer.next )

File: src/sorting/python/heap_sort/README.md [hunk 6]
```diff
@@ -1,4 +1,4 @@
-Merge sort implementation
+Heap sort implementation
 Author: Vlad421
 
 Implementation of heap sort algorithm
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/README.md:4]: Well this is definately should not be here!!
You made corrections to heap_sort staying in the linked_list branch!!

File: src/sorting/python/heap_sort/src/heap_sort.py [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
 """
-__author__ = "Vlad421"
 
 
 def heap_sort(array_to_sort: list[any], is_ascending: bool = True) -> list[any]:
-    """Sorts a list using heap_sort algorithm, mutate given list
+    """Sorts a list using heap_sort algorithm, returns sorted List
 
     Parameters:
         array_to_sort: list[any] - List of items to sort
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/src/heap_sort.py:17]: the same

File: src/sorting/python/heap_sort/tests/test_heap_sort.py [hunk 8]
```diff
@@ -7,6 +6,7 @@
 file_dir = os.path.dirname(os.path.dirname(__file__))
 sys.path.append(file_dir + "/src")
 
+from heap_sort import heap_sort
 
 class TestHeapSortV1(unittest.TestCase):
     def test_empty_array(self):
```
Comment (tvsirius) [inline src/sorting/python/heap_sort/tests/test_heap_sort.py:12]: the same

**Review by tvsirius — CHANGES_REQUESTED** [review]
Hi, all looks good, with small comments
But there is an error in .remove()
you can add this test to see it:
```
    def test_remove_error_check(self):
        linked_list = LinkedList()
        linked_list.push_back(1)
        linked_list.push_back(2)
        linked_list.push_back(3)
        linked_list.remove(2)
        self.assertEqual(linked_list._LinkedList__data[0].value, 1)
        self.assertEqual(linked_list._LinkedList__data[0].next.value, 3)
```
FAIL: test_remove_error_check (__main__.TestLinkedList.test_remove_error_check)
----------------------------------------------------------------------
Traceback (most recent call last):
in test_remove_error_check
    self.assertEqual(linked_list._LinkedList__data[0].next.value, 3)
AssertionError: 2 != 3


Needs correction

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 3 above)
Comment (delvor) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:64]: Up to implementation I think. I wouldn't call it pop. For me pop does not speak for itself about its purpose

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (delvor) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: It swaps Node references while removing. If I move to 103 it would be wrong node reference

File: src/sorting/python/heap_sort/README.md (hunk 6 above)
Comment (delvor) [reply src/sorting/python/heap_sort/README.md:4]: All correct. It fix to the previous merge request. I write it in a commit message. Look at the file changed.

File: src/sorting/python/heap_sort/src/heap_sort.py (hunk 7 above)
Comment (delvor) [reply src/sorting/python/heap_sort/src/heap_sort.py:17]: Fix to previous pull request

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 2 above)
Comment (delvor) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:52]: It's done on purpose. Turns out everything in Python is object. Wasn't sure if the return statement returns value or reference.

**Comment by delvor** [comment]
You forgot the dot between LinkedList and __data
Should be like this:
>self.assertEqual(linked_list._LinkedList.__data[0].value

Also you shouldn't be able to access the __data variable it's private. But who knows. Python has its flaws

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (tvsirius) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: Check the test proving it does wrong

**Comment by tvsirius** [comment]
python kind of like 'protect' marked with __ __variables, 
but you still can access them with this constrution
        self.assertEqual(linked_list._LinkedList__data[0].value, 1)
        self.assertEqual(linked_list._LinkedList__data[0].next.value, 3)
        
run this code


MyClass.__protected - will give no access
MyClass._MyClass__proctected - will give access

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (tvsirius) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: Yes, but in the picture they swap after changing Pos and you swap before

**Comment by delvor** [comment]
> python kind of like 'protect' marked with __ __variables, but you still can access them with this constrution self.assertEqual(linked_list._LinkedList__data[0].value, 1) self.assertEqual(linked_list._LinkedList__data[0].next.value, 3)
> 
> run this code
> 
> MyClass.__protected - will give no access MyClass._MyClass__proctected - will give access

Indeed. You can access private vars through this construction. 

Fixed issue.

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (delvor) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: test is failed not because of it. look at the changes

File: src/collections_and_containers/python/linked_list/src/linked_list.py [hunk 9]
```diff
@@ -87,15 +87,16 @@ def remove(self, element: any):
         if not self.empty():
 
             counter: int = 0
-            pointer: self.Node = self.__data[counter]
-            
+            pointer: self.Node = self.__data[counter].next
```
Comment (tvsirius) [inline src/collections_and_containers/python/linked_list/src/linked_list.py:90]: Yes, now you move the pointer to next

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (tvsirius) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: Yep, now it works!

**Review by tvsirius — APPROVED** [review]
Good

File: src/collections_and_containers/python/linked_list/src/linked_list.py (hunk 5 above)
Comment (delvor) [reply src/collections_and_containers/python/linked_list/src/linked_list.py:100]: Nah. Could fix the error just changed expression by adding .next at the end. 

Anyway. Provided solution works 😀
